src/common/contracts/ao-process.ts

```diff
@@ -60,72 +66,83 @@ export class AOProcess implements AOContract {
     retries?: number;
     fromAddress?: string;
   }): Promise<K> {
+    this.logger.debug(`Evaluating read interaction on process`, {
+      tags,
+      processId: this.processId,
+    });
+    // map tags to inputs
+    const dryRunInput = {
+      process: this.processId,
+      tags,
+    };
+    if (fromAddress !== undefined) {
+      dryRunInput['Owner'] = fromAddress;
+    }
+
     let attempts = 0;
-    let lastError: Error | undefined;
+    let result: DryRunResult | undefined = undefined;
+
     while (attempts < retries) {
       try {
-        this.logger.debug(`Evaluating read interaction on process`, {
-          tags,
-          processId: this.processId,
-        });
-        // map tags to inputs
-        const dryRunInput = {
-          process: this.processId,
-          tags,
-        };
-        if (fromAddress !== undefined) {
-          dryRunInput['Owner'] = fromAddress;
-        }
-        const result = await this.ao.dryrun(dryRunInput);
-        this.logger.debug(`Read interaction result`, {
-          result,
-          processId: this.processId,
-        });
-
-        const error = errorMessageFromOutput(result);
-        if (error !== undefined) {
-          throw new Error(error);
-        }
-
-        if (result.Messages === undefined || result.Messages.length === 0) {
-          this.logger.debug(
-            `Process ${this.processId} does not support provided action.`,
-            {
-              result,
-              tags,
-              processId: this.processId,
-            },
-          );
-          throw new Error(
-            `Process ${this.processId} does not support provided action.`,
-          );
-        }
-        const messageData = result.Messages?.[0]?.Data;
-
-        // return undefined if no data is returned
-        if (this.isMessageDataEmpty(messageData)) {
-          return undefined as K;
-        }
-
-        const response: K = safeDecode<K>(messageData);
-        return response;
-      } catch (error: any) {
+        result = await this.ao.dryrun(dryRunInput);
+        // break on successful return of result
```

Comment:
We still want to check the result for the error and construct an error to throw it i think.
```
 const error = errorMessageFromOutput(result);
        if (error !== undefined) {
          throw new Error(error);
        }
```